plutus-tx-plugin/src/PlutusTx/Compiler/Type.hs

```diff
@@ -118,32 +118,34 @@ compileTyCon tc
     | otherwise = do
 
     let tcName = GHC.getName tc
+        lexName = LexName tcName
     whenM (blackholed tcName) $ throwSd UnsupportedError $ "Recursive newtypes, use data:" GHC.<+> GHC.ppr tcName
-    maybeDef <- PIR.lookupType () (LexName tcName)
+    -- See Note [Dependency tracking]
+    modifyCurDeps (\d -> Set.insert lexName d)
+    maybeDef <- PIR.lookupType () lexName
     case maybeDef of
         Just ty -> pure ty
-        Nothing -> do
-            dcs <- getDataCons tc
-            usedTcs <- getUsedTcs tc
-            let deps = fmap GHC.getName usedTcs ++ fmap GHC.getName dcs
-
+        -- See Note [Dependency tracking]
+        Nothing -> withCurDef lexName $ do
             tvd <- compileTcTyVarFresh tc
-
             case GHC.unwrapNewTyCon_maybe tc of
                 Just (_, underlying, _) -> do
                     -- See Note [Coercions and newtypes]
                     -- See Note [Occurrences of recursive names]
+                    -- We do this for dependency tracking, we won't use it due to the blackholing
+                    PIR.defineType lexName (PIR.Def tvd (PIR.mkTyVar () tvd)) mempty
                     -- Type variables are in scope for the rhs of the alias
                     alias <- mkIterTyLamScoped (GHC.tyConTyVars tc) $ blackhole (GHC.getName tc) $ compileTypeNorm underlying
-                    PIR.defineType (LexName tcName) (PIR.Def tvd alias) (Set.fromList $ LexName <$> deps)
-                    PIR.recordAlias @LexName @uni @fun @() (LexName tcName)
+                    PIR.modifyTypeDef lexName (const $ PIR.Def tvd alias)
```

Comment:
Is "define and then modify" the same as "define the resulting type"?

Reply:
Hmm, I guess I need to explain the dance more carefully. What we do here is define it with a "fake" definition, then compile the real definition, during the course of which we will discover dependencies (adding them to the original fake definition), and then at the end we modify the definition to the final one. I'll add this basic pattern to the Note.